File: mini-projects/csis2260_operating_system/metrics_layer_strategy/metrics_layer.py
from collections import deque
from dataclasses import dataclass
from typing import Deque, Protocol, Dict

@dataclass
class Process:
    pid: str
    remaining: int
    priority: int = 0

class SchedulingStrategy(Protocol):
    def dequeue_next(self, ready: Deque[Process]) -> Process: ...

class Fcfs:
    def dequeue_next(self, ready: Deque[Process]) -> Process:
        return ready.popleft()
    
class Sjf:
    def dequeue_next(self, ready: Deque[Process]) -> Process:
        best = None
        for _ in range(len(ready)):
            p = ready.popleft()
            if best is None or p.remaining < best.remaining:
                if best is not None:
                    ready.append(best)
                best = p
            else:
                ready.append(p)
        return best

class PrioritySched:
    def dequeue_next(self, ready: Deque[Process]) -> Process:
        # lower priority value first
        best = None
        for _ in range(len(ready)):
            p = ready.popleft()
            if best is None or p.priority < best.priority:
                if best is not None:
                    ready.append(best)
                best = p
            else:
                ready.append(p)
        return best
    
class Metrics:
    def __init__(self):
        self.context_switches = 0
        self.first_response: Dict[str, int] = {}
        self.finish_time: Dict[str, int] = {}
        self.burst_seen: Dict[str, bool] = {}
        self.now = 0

    def on_dispatch(self, p: Process):
        self.context_switches += 1
        if p.pid not in self.burst_seen:
            self.first_response[p.pid] = self.now
            self.burst_seen[p.pid] = True

    def on_execute(self, run: int):
        self.now += run
    
    def on_complete(self, p: Process):
        self.finish_time[p.pid] = self.now

    def summary(self, original_bursts: Dict[str, int]):
        res = {}
        for pid, total in original_bursts.items():
            ft = self.finish_time[pid]
            rt = self.first_response.get(pid, 0)
            tat = ft # arrival=0
            wt = tat - total
            res[pid] = {"response": rt, "turnaround": tat, "waiting": wt}
        return {
            "per_process": res,
            "context_switches": self.context_switches,
            "makespan": self.now,
        }
    
class Scheduler:
    def __init__(self, ready: Deque[Process], strategy: SchedulingStrategy, quantum: int):
        assert quantum > 0
        self.ready = ready
        self.strategy = strategy
        self.quantum = quantum
        self.metrics = Metrics()
        self.original = {p.pid: p.remaining for p in list(ready)}

    def run(self):
        while self.ready:
            p = self.strategy.dequeue_next(self.ready)
            self.metrics.on_dispatch(p)
            run = min(self.quantum, p.remaining)
            p.remaining -= run
            self.metrics.on_execute(run)
            if p.remaining > 0:
                self.ready.append(p)
            else:
                self.metrics.on_complete(p)
        return self.metrics.summary(self.original)


if __name__ == "__main__":
    base = deque([
        Process("P1", 5, priority=2),
        Process("P2", 3, priority=1),
        Process("P3", 8, priority=3),
    ])
    print("FCFS:", Scheduler(deque(base), Fcfs(), 10**9).run())
    print("SJF:", Scheduler(deque(base), Sjf(), 10**9).run())
    print("PRIO:", Scheduler(deque(base), PrioritySched(), 10**9).run())
    print("RR q=2:", Scheduler(deque(base), Fcfs(), 2).run())

File: mini-projects/csis2260_operating_system/metrics_layer_strategy/metrics_layer.js
class Process {
  constructor(pid, remaining, priority = 0) {
    this.pid = pid;
    this.remaining = remaining;
    this.priority = priority;
  }
}

class Metrics {
  constructor() {
    this.contextSwitches = 0;
    this.firstResponse = {};
    this.finishTime = {};
    this.burstSeen = {};
    this.now = 0;
  }

  OnDispatch(p) {
    this.contextSwitches++;
    if (!(p.pid in this.burstSeen)) {
      this.firstResponse[p.pid] = this.now;
      this.burstSeen[p.pid] = true;
    }
  }

  onExecute(run) {
    this.now += run;
  }

  onComplete(p) {
    this.finishTime[p.pid] = this.now;
  }

  Summary(original) {
    const per = {};
    for (const [pid, total] of Object.entries(original)) {
      const ft = this.finishTime[pid];
      const rt = this.firstResponse[pid];
      const tat = ft;
      const wt = tat - total;
      per[pid] = { response: rt, turnaround: tat, waiting: wt };
    }
    return {
      per_process: per,
      context_switches: this.contextSwitches,
      makespan: this.now,
    };
  }
}

class FcfsStrategy {
  dequeueNext(q) {
    return q.shift();
  }
}

class SjfStrategy {
  dequeueNext(q) {
    let best = 0;
    for (let i = 1; i < q.length; i++) {
      if (q[i].remaining < q[best].remaining) best = i;
    }
    return q.splice(best, 1)[0];
  }
}

class PriorityStrategy {
  dequeueNext(q) {
    let best = 0;
    for (let i = 1; i < q.length; i++) {
      if (q[i].priority < q[best].priority) best = i;
    }
    return q.splice(best, 1)[0];
  }
}

class Scheduler {
  constructor(ready, strategy, quantum) {
    this.ready = [...ready];
    this.strategy = strategy;
    this.quantum = quantum;
    this.metrics = new Metrics();
    this.original = Object.fromEntries(
      this.ready.map((p) => [p.pid, p.remaining])
    );
  }
  run() {
    while (this.ready.length > 0) {
      const p = this.strategy.dequeueNext(this.ready);
      this.metrics.OnDispatch(p);
      const run = Math.min(this.quantum, p.remaining);
      p.remaining -= run;
      this.metrics.onExecute(run);
      if (p.remaining > 0) {
        this.ready.push(p);
      } else {
        this.metrics.onComplete(p);
      }
    }
    return this.metrics.Summary(this.original);
  }
}

const base1 = [
  new Process("P1", 5, 2),
  new Process("P2", 3, 1),
  new Process("P3", 8, 3),
];
console.log(
  "FCFS:",
  new Scheduler(base1, new FcfsStrategy(), Number.MAX_SAFE_INTEGER).run()
);

const base2 = [
  new Process("P1", 5, 2),
  new Process("P2", 3, 1),
  new Process("P3", 8, 3),
];
console.log(
  "SJF:",
  new Scheduler(base2, new SjfStrategy(), Number.MAX_SAFE_INTEGER).run()
);

const base3 = [
  new Process("P1", 5, 2),
  new Process("P2", 3, 1),
  new Process("P3", 8, 3),
];
console.log(
  "PRIO:",
  new Scheduler(base3, new PriorityStrategy(), Number.MAX_SAFE_INTEGER).run()
);
const base4 = [
  new Process("P1", 5, 2),
  new Process("P2", 3, 1),
  new Process("P3", 8, 3),
];
console.log("RR q=2:", new Scheduler(base4, new FcfsStrategy(), 2).run());
